### PR TITLE
Submodule latest PyTorch and remove ATen submodule

### LINKED ISSUE #155: Bump ATen submodule
1. required for supporting pytorch master
2. required to unblock @salexspb  and team

### LINKED ISSUE #328: ATen submodule points to non-existing commit
Running `git clone https://github.com/facebookresearch/TensorComprehensions.git --recursive` gives
```
error: Server does not allow request for unadvertised object f234e94aedffdbd9a47092ac36ea1784da8bc50a
Fetched in submodule path 'third-party/ATen', but it did not contain f234e94aedffdbd9a47092ac36ea1784da8bc50a. Direct fetching of that commit failed.
```
and `git submodule update --init --recursive` gives
```
error: Server does not allow request for unadvertised object f234e94aedffdbd9a47092ac36ea1784da8bc50a
Fetched in submodule path 'third-party/ATen', but it did not contain f234e94aedffdbd9a47092ac36ea1784da8bc50a. Direct fetching of that commit failed.
```
It looks like the current submodule points to https://github.com/zdevito/aten/tree/f234e94aedffdbd9a47092ac36ea1784da8bc50a, but that commit has disappeared (rewrite/force push?). It seems like it was replaced with https://github.com/zdevito/ATen/tree/13d8a3f2d43ccd7d274b66b5f68372ccbd16de17, which doesn't compile since it looks for `CMakeLists.txt` which is no longer there...

I'm trying to compile now by checking out https://github.com/zdevito/aten/tree/old_master, but could someone confirm what the correct ATen commit to compile with is?

### PR DESCRIPTION
**Pull request by prigoyal**
This PR makes following changes:
1. Bump ATen submodule to latest master
2. use pytorch-nightly conda package for python tests - these conda packages are built for pytorch master on daily basis. This will also allow us to track aten master and bump the submodule when API changes happen in ATen 
3. submodule pytorch to get aten and remove the aten submodule. I'll move to using caffe2 from pytorch in a later PR.
4. build and pushed new docker images to TC dockerhub - CircleCI tests pass.
5. enable C++ test as well on cuda9-conda environment build - it used to be python test only in cuda9-conda env.
6. TC codebase changes to have updated ATen api calls.

closes #155 
closes #328

### CODE REVIEW

**Comment by prigoyal** [comment]
@mingzhe09088 , @salexspb this PR should unblock the compilation in fbcode. The only blocked for this PR to merge is refreshing the pytorch conda package. I am building a temporary conda package now and will update PR :)

also cc @apaszke for any thoughts/review :)

File: examples/example_group_convolution.cc [hunk 1]
```diff
@@ -340,7 +340,7 @@ TEST_F(GroupConvolution, ATenGroupConvolutionReference) {
           auto weight_g = subtensor(W2, 0, G, g);
           auto bias_g = subtensor(B1, 0, G, g);
           outputs[g] =
-              at::conv2d(input_g, weight_g, at::IntList({KH, KW}), bias_g);
+              at::thnn_conv2d(input_g, weight_g, at::IntList({KH, KW}), bias_g);
```
Comment (apaszke) [inline examples/example_group_convolution.cc]: There’s no reason to use the thnn impl. Just remove the IntList arg

Comment (prigoyal) [reply examples/example_group_convolution.cc]: cool, thanks @apaszke , will do so.

Comment (prigoyal) [reply examples/example_group_convolution.cc]: @apaszke , it looks like `conv2d` doesn't take kernel size as argument https://github.com/zdevito/ATen/blob/d03c3e37b62827865ba210b31528d9e00024cf7f/src/ATen/native/Convolution.cpp#L242-L247 
but `thnn_conv2d` does https://github.com/zdevito/ATen/blob/85f56d1adce2ed0b1f7859db399bfc2e8c26a461/src/ATen/nn.yaml#L271-L273 . do you know how `conv2d` handles the kernel size argument?

Comment (apaszke) [reply examples/example_group_convolution.cc]: It's redundant, so it is inferred from the weights.

**Comment by apaszke** [comment]
@prigoyal looks like the CI is failing because some of the boxes have MKL installed, but don't have the `mkl-includes` package which is an extra dependency (if MKL is found) now.

**Comment by prigoyal** [comment]
hey @apaszke , there are more reasons https://ci.pytorch.org/jenkins/job/tensorcomp-builds/job/linux-trusty-gcc4.8-cuda8-cudnn7-py3-conda-build/42/console the pytorch conda package is not compatible so we need a refreshed package. 

the mkl issue is also happening on the 1404 circle box which is again conda env and installs pytorch.

**Comment by mingzhe09088** [comment]
I saw the following compilation issues when building this bump-aten branch locally. Not sure why CI build passed: 
In file included from /home/mingzhe0908/TC/bump-aten/test/test_mapper_llvm.cc:32:0:
/home/mingzhe0908/TC/bump-aten/test/test_harness_aten.h: In function ‘bool checkRtol(const at::Tensor&, std::vector<at::Tensor>, double, double)’:
/home/mingzhe0908/TC/bump-aten/test/test_harness_aten.h:43:40: error: ‘class at::Scalar’ has no member named ‘toCFloat’
     maxValue = fmax(tensor.abs().max().toCFloat(), maxValue);
                                        ^
/home/mingzhe0908/TC/bump-aten/test/test_harness_aten.h:45:35: error: ‘class at::Scalar’ has no member named ‘toCFloat’
   auto maxDiff = diff.abs().max().toCFloat();
                                   ^
/home/mingzhe0908/TC/bump-aten/test/test_mapper_llvm.cc: In member function ‘virtual void LLVMCodegen_Convolution_Test::TestBody()’:
/home/mingzhe0908/TC/bump-aten/test/test_mapper_llvm.cc:213:44: error: could not convert ‘B’ from ‘at::Tensor’ to ‘at::IntList {aka at::ArrayRef<long int>}’
   at::Tensor expected = at::conv2d(I, W1, B);
                                            ^

**Comment by prigoyal** [comment]
Hi Mingzhe,

CI has not passed. It’s failing build for exact same reasons.

**Comment by prigoyal** [comment]
hello everyone, just a quick update: right now @soumith is looking into the nightly for pytorch and providing us with the pytorch package. That should unblock this PR. Thanks :)

**Comment by mingzhe09088** [comment]
@prigoyal Do we have any update here? Once this is merged, I can start sync master to fbcode.

**Comment by prigoyal** [comment]
Hi @mingzhe09088 , we are waiting for pytorch nightly to happen so we can have conda package for it and pass CI. I am adding you to the messenger group so you can track it :)

**Comment by prigoyal** [comment]
@salexspb please feel free to review this PR. Once this PR is merged, I will have to trigger docker image build on Jenkins CI manually. It gets tricky to build docker image for a PR on Jenkins current setup. Thanks :)

File: test/cuda/test_tc_mapper.cc [hunk 2]
```diff
@@ -355,7 +355,7 @@ def fun(float(B, R) LUT, int32(B, N) I) -> (O) {
 ///////////////////////////////////////////////////////////////////////////////
 // SpatialBatchNormalization
 ///////////////////////////////////////////////////////////////////////////////
-TEST_F(TcCudaMapperTest, SpatialBatchNormalization) {
+TEST_F(TcCudaMapperTest, DISABLED_SpatialBatchNormalization) {
```
Comment (nicolasvasilache) [inline test/cuda/test_tc_mapper.cc]: why disable the test?

File: cmake/Modules_CUDA_fix/FindCUDA.cmake [hunk 3]
```diff
@@ -0,0 +1,2004 @@
...
+#                            Windows only.
+#   CUDA_nvcuvid_LIBRARY  -- CUDA Video Decoder library.
+#                            Only available for CUDA version 3.2+.
+#                            Windows only.
+#
+
+#   James Bigler, NVIDIA Corp (nvidia.com - jbigler)
+#   Abe Stephens, SCI Institute -- http://www.sci.utah.edu/~abe/FindCuda.html
+#
+#   Copyright (c) 2008 - 2009 NVIDIA Corporation.  All rights reserved.
```
Comment (ftynse) [inline cmake/Modules_CUDA_fix/FindCUDA.cmake]: Do we want an external copyright in our codebase?

File: cmake/Modules_CUDA_fix/FindCUDA.cmake [hunk 4]
```diff
@@ -0,0 +1,2004 @@
...
+#
+#   Copyright (c) 2007-2009
+#   Scientific Computing and Imaging Institute, University of Utah
+#
+#   This code is licensed under the MIT License.  See the FindCUDA.cmake script
+#   for the text of the license.
+
+# The MIT License
```
Comment (ftynse) [inline cmake/Modules_CUDA_fix/FindCUDA.cmake]: Do we want a different license in our codebase?

File: cmake/Modules_CUDA_fix/FindCUDA.cmake (hunk 3 above)
Comment (prigoyal) [reply cmake/Modules_CUDA_fix/FindCUDA.cmake]: I am simply copy-pasting the files from PyTorch repo which is BSD license itself: https://github.com/pytorch/pytorch/tree/master/cmake/Modules_CUDA_fix 

I am not sure about the concern. In my understanding, explicitly saying the copyright in the file means that it applies to that file and not the entire codebase.

File: cmake/Modules_CUDA_fix/FindCUDA.cmake (hunk 4 above)
Comment (prigoyal) [reply cmake/Modules_CUDA_fix/FindCUDA.cmake]: I am not sure about the concern. In my understanding, explicitly saying the copyright in the file means that it applies to that file and not the entire codebase.

File: cmake/Modules_CUDA_fix/FindCUDA.cmake (hunk 3 above)
Comment (ftynse) [reply cmake/Modules_CUDA_fix/FindCUDA.cmake]: There was some reason why we could not reuse some polyhedral code with external copyright, and why CUB went into a submodule.  If the policy has changed, I'd like to know. 

cc @nicolasvasilache

Comment (prigoyal) [reply cmake/Modules_CUDA_fix/FindCUDA.cmake]: @salexspb, based on your experience with C2, do you have some info about how licensing works with CLA?

File: test/cuda/test_tc_mapper.cc (hunk 2 above)
Comment (prigoyal) [reply test/cuda/test_tc_mapper.cc]: it doesn't work, it throws dimension errors (I corrected that), very big output errors (not sure why). Since batchnorm is not easy to debug, I disabled it.

File: cmake/Modules_CUDA_fix/FindCUDA.cmake (hunk 3 above)
Comment (ftynse) [reply cmake/Modules_CUDA_fix/FindCUDA.cmake]: Let me put this differently: can I also put new files with external copyright (mine) and a compatible license?  If so, why we were not doing this from the start?

**Comment by prigoyal** [comment]
@salexspb , I have tried to recycle those files from caffe2 repo. The reason for adding the cmake files are in this issue: pytorch/pytorch#6556

can you take a look and provide feedback on the PR now? Thanks.

**Comment by salexspb** [comment]
Reusing these files from Caffe2 is actually not great due to caffe2 being a submodule which is no longer possible. Caffe2 will probably stay as a dead code in this repo, we will need to figure out another way to test against it.  As it is now just a folder in pytoch repo.

 I would add those files directly if they are needed in any case. Still I am going to accept this diff as it doesn't introduce any new copyrights and helps with aten issues. And @nicolasvasilache was going to follow up with OSS team about copyrights.

File: docker/common/install_pip_python.sh [hunk 5]
```diff
@@ -4,10 +4,23 @@ set -ex
...
 # Install common dependencies
 apt-get update
-apt-get install -y --no-install-recommends python3-dev python3-pip python3-setuptools
+apt-get install -y --no-install-recommends python3-dev python3-setuptools
 
-pip3 install --upgrade pip
-pip3 install numpy decorator six future setuptools pyyaml
+# Install pip from source. The python-pip package on Ubuntu Trusty is old
+# and upon install numpy doesn't use the binary distribution, and fails to compile it from source.
+# curl -O https://pypi.python.org/packages/11/b6/abcb525026a4be042b486df43905d6893fb04f05aac21c32c638e939e447/pip-9.0.1.tar.gz
```
Comment (ftynse) [inline docker/common/install_pip_python.sh]: commented-out code

File: test/cuda/test_tc_mapper.cc [hunk 6]
```diff
@@ -355,7 +355,8 @@ def fun(float(B, R) LUT, int32(B, N) I) -> (O) {
 ///////////////////////////////////////////////////////////////////////////////
 // SpatialBatchNormalization
 ///////////////////////////////////////////////////////////////////////////////
-TEST_F(TcCudaMapperTest, SpatialBatchNormalization) {
+// TODO: xxx https://github.com/facebookresearch/TensorComprehensions/issues/319
```
Comment (ftynse) [inline test/cuda/test_tc_mapper.cc]: "xxx" is not a very useful comment

File: cmake/CMakeInitializeConfigs.cmake [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
+# file Copyright.txt or https://cmake.org/licensing for details.
```
[the remark this answers is not in the document]
Comment (prigoyal) [reply cmake/CMakeInitializeConfigs.cmake]: https://cmake.org/licensing/ link mentioned after "or" directs to the Copyright.txt file. I have simply copied this file from here https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/CMakeInitializeConfigs.cmake without any modifications. Thanks.

[the remark this answers is not in the document]
Comment (prigoyal) [reply cmake/CMakeInitializeConfigs.cmake]: update: I have reached out to our legal team for this as well.

**Comment by prigoyal** [comment]
@salexspb , yep those are very valid points. We should actually move to submodule pytorch repo and that will give us aten, caffe2, pytorch all. what do you think? This is a separate item to do on my list.

File: cmake/CMakeInitializeConfigs.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (ftynse) [reply cmake/CMakeInitializeConfigs.cmake]: In such cases, it is better to avoid confusion and have legal clearance.  (1) I don't know if `Copyright.txt` file actually _requires_ you to copy itself along with the code. It is often the case with open source licenses: failing to provide the copy of the _license_ file violates the license.  (2) There is a `COPYRIGHT` file in the root directory which has a _different_ copyright and license, and may be confused with what is mentioned here.  (3) If there is a way to add code with external copyright, I want my code to have my copyright ;)

[the remark this answers is not in the document]
Comment (ftynse) [reply cmake/CMakeInitializeConfigs.cmake]: This is the first point of the [Copyright.txt](https://gitlab.kitware.com/cmake/cmake/blob/master/Copyright.txt) file you did not copy
```
CMake - Cross Platform Makefile Generator
Copyright 2000-2018 Kitware, Inc. and Contributors
All rights reserved.

* Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
```

I don't see the "above copyright notice" nor the "following disclaimer" in their code you essentially redistribute.  I would claim this violates the license.

**Comment by ftynse** [comment]
> We should actually move to submodule pytorch repo and that will give us aten, caffe2, pytorch all. what do you think?

This should be discussed in a general meeting, it was mentioned today.
I would prefer the submoduling to go the other way around.

**Comment by ftynse** [comment]
Moving discussion away from github actually looks like **de**centralization ;)

I don't really have time for more writing at this moment, that's why I would have preferred this issue to be discussed in a general meeting (for which I have a fixed allocated time slot).

**Comment by skimo-openhub** [comment]
Also, the one-line summary of 26222c43115bf5fd972ab8c8bb1d5e7a43a35762 is wrong.
It doesn't look like you are removing a pytorch submodule.

**Comment by prigoyal** [comment]
@salexspb, if you would like to review this PR, this is ready to be merged. Jenkins CI will keep failing since it requires new docker images that are triggered to build on Jenkins CI itself after the PR is merged.

File: .circleci/config.yml [hunk 8]
```diff
@@ -4,7 +4,7 @@ jobs:
     working_directory: ~/TensorComprehensions
     resource_class: xlarge
     docker:
-      - image: tensorcomprehensions/linux-trusty-gcc4.8-cuda8-cudnn6-py3-conda:x86
+      - image: tensorcomprehensions/linux-trusty-gcc4.9-cuda8-cudnn7-py3-conda:1
```
Comment (nicolasvasilache) [inline .circleci/config.yml]: Does this bump the required gcc from 4.8.5 to 4.9 ?
If so, please document and explain why.

Comment (prigoyal) [reply .circleci/config.yml]: reason is in the commit message. Where else do you recommend to document this?
